File: FundComponents/ImpactDiarrhoeaComponent.cs
// FUND - Climate Framework for Uncertainty, Negotiation and Distribution
// Copyright (C) 2012 David Anthoff and Richard S.J. Tol
// http://www.fund-model.org
// Licensed under the MIT license
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Esmf;
using Fund.CommonDimensions;

namespace Fund.Components
{

    public interface IImpactDiarrhoeaState
    {
        IVariable2Dimensional<Timestep, Region, Double> diadead { get; }
        IVariable2Dimensional<Timestep, Region, Double> diasick { get; }

        IParameter1Dimensional<Region, Double> diamort { get; }
        Double diamortel { get; }
        Double diamortnl { get; }

        IParameter1Dimensional<Region, Double> diayld { get; }
        Double diayldel { get; }
        Double diayldnl { get; }

        IParameter2Dimensional<Timestep, Region, Double> income { get; }
        IParameter2Dimensional<Timestep, Region, Double> population { get; }
        IParameter1Dimensional<Region, Double> gdp90 { get; }
        IParameter1Dimensional<Region, Double> pop90 { get; }

        IParameter1Dimensional<Region, Double> temp90 { get; }
        IParameter1Dimensional<Region, Double> bregtmp { get; }
        IParameter2Dimensional<Timestep, Region, Double> regtmp { get; }
    }

    public class ImpactDiarrhoeaComponent
    {
        public void Run(Clock clock, IImpactDiarrhoeaState state, IDimensions dimensions)
        {
            var s = state;
            var t = clock.Current;

            foreach (var r in dimensions.GetValues<Region>())
            {
                var ypc = 1000.0 * s.income[t, r] / s.population[t, r];
                var ypc90 = 1000.0 * s.gdp90[r] / s.pop90[r];

                // 0.49 is the increase in global temperature from pre-industrial to 1990
                var absoluteRegionalTempPreIndustrial = s.temp90[r] - 0.49 * s.bregtmp[r];

                if (absoluteRegionalTempPreIndustrial > 0.0)
                {
                    s.diadead[t, r] = s.diamort[r] * s.population[t, r] * Math.Pow(ypc / ypc90, s.diamortel)
                        * (Math.Pow((absoluteRegionalTempPreIndustrial + s.regtmp[t, r]) / absoluteRegionalTempPreIndustrial, s.diamortnl) - 1.0);

                    s.diasick[t, r] = s.diayld[r] * s.population[t, r] * Math.Pow(ypc / ypc90, s.diayldel)
                        * (Math.Pow((absoluteRegionalTempPreIndustrial + s.regtmp[t, r]) / absoluteRegionalTempPreIndustrial, s.diayldnl) - 1.0);
                }
                else
                {
                    s.diadead[t, r] = 0.0;
                    s.diasick[t, r] = 0.0;
                }
            }
        }

    }
}

File: src/components/ImpactDiarrhoeaComponent.jl
@defcomp impactdiarrhoea begin
    regions = Index()

    diadead = Variable(index=[time,regions])
    diasick = Variable(index=[time,regions])

    diamort = Parameter(index=[regions])
    diamortel = Parameter(default=-1.578625)
    diamortnl = Parameter(default=1.141714)

    diayld = Parameter(index=[regions])
    diayldel = Parameter(default=-0.418406)
    diayldnl = Parameter(default=0.699241)

    income = Parameter(index=[time,regions])
    population = Parameter(index=[time,regions])
    gdp90 = Parameter(index=[regions])
    pop90 = Parameter(index=[regions])

    temp90 = Parameter(index=[regions])
    bregtmp = Parameter(index=[regions])
    regtmp = Parameter(index=[time,regions])

    function run_timestep(p, v, d, t)

        for r in d.regions
            ypc = 1000.0 * p.income[t, r] / p.population[t, r]
            ypc90 = 1000.0 * p.gdp90[r] / p.pop90[r]

            # 0.49 is the increase in global temperature from pre-industrial to 1990
            absoluteRegionalTempPreIndustrial = p.temp90[r] - 0.49 * p.bregtmp[r]

            if absoluteRegionalTempPreIndustrial > 0.0
                v.diadead[t, r] = p.diamort[r] * p.population[t, r] * (ypc / ypc90)^p.diamortel * (((absoluteRegionalTempPreIndustrial + p.regtmp[t, r]) / absoluteRegionalTempPreIndustrial)^p.diamortnl - 1.0)

                v.diasick[t, r] = p.diayld[r] * p.population[t, r] * (ypc / ypc90)^p.diayldel * (((absoluteRegionalTempPreIndustrial + p.regtmp[t, r]) / absoluteRegionalTempPreIndustrial)^p.diayldnl - 1.0)
            else
                v.diadead[t, r] = 0.0
                v.diasick[t, r] = 0.0
            end
        end
    end
end